### PR TITLE
(RK-246) Env-level purging and configuration

### PR DESCRIPTION
**Pull request by scotje**
Still needs a few more unit tests for the new settings, but can start being reviewed at this point I think.

### CODE REVIEW

File: doc/dynamic-environments/configuration.mkd [hunk 1]
```diff
@@ -200,12 +200,81 @@ sources:
...
+```yaml
+---
+deploy:
+  purge_levels: [ 'deployment', 'environment', 'puppetfile' ]
+```
+
+This setting currently only impacts the "deploy environment" action.
+
+To maintain parity with existing behavior before this setting was added, the
+default value is `['deployment', 'puppetfile']`.
```
Comment (andersonmills) [inline doc/dynamic-environments/configuration.mkd]: How about reverse this sentence?
The default value is `[dep pup]`, which maintains parity...
Much easier to see the default.

File: doc/dynamic-environments/configuration.mkd [hunk 2]
```diff
@@ -200,12 +200,81 @@ sources:
...
+The `deploy` setting is a top level setting for controlling how r10k deploys
 behave. At this point only new settings are included under this setting, but in
 the long term the current top level deploy settings will be moved under
 `deploy`.
 
-#### write_lock
+#### purge\_levels
+
+The `purge_levels` setting controls how aggressively r10k will purge unmanaged
+content during a deployment. Must be a list of values indicating at what levels
```
Comment (andersonmills) [inline doc/dynamic-environments/configuration.mkd]: The "Must" sentence is not a complete sentence.

File: doc/dynamic-environments/configuration.mkd [hunk 3]
```diff
@@ -200,12 +200,81 @@ sources:
...
+This setting currently only impacts the "deploy environment" action.
+
+To maintain parity with existing behavior before this setting was added, the
+default value is `['deployment', 'puppetfile']`.
+
+The effect of enabling the various purge levels is as follows:
+
+##### deployment
+
+After each deploy, in the configured basedir, recursively remove any content
```
Comment (andersonmills) [inline doc/dynamic-environments/configuration.mkd]: I don't love writing a definition to look like a sentence when it is not one, so I think the non-sentence or the format should change. I guess this is maybe an imperative sentence where the implied "you" is r10k?

File: doc/dynamic-environments/configuration.mkd [hunk 4]
```diff
@@ -200,12 +200,81 @@ sources:
...
+found which is not managed by one of the sources declared in your r10k.yaml
+configuration. Note that disabling this level of purging may cause the number of
+deployed environments to grow without bound; deleting branches from your control
+repo would no longer cause the matching environment to be purged.
+
+##### environment
+
+When a given environment is deployed, recursively remove any content found
```
Comment (andersonmills) [inline doc/dynamic-environments/configuration.mkd]: non-sentence or maybe imperative?

File: doc/dynamic-environments/configuration.mkd [hunk 5]
```diff
@@ -200,12 +200,81 @@ sources:
...
+which is not either committed to the control repo branch that maps to that
+environment, or declared in a Puppetfile committed to that branch. Note that
+the .r10k-deploy.json file is exempt from this purging. Additionally, no
+content will be purged if an error is encountered while deploying the
+environment's Puppetfile content.
+
+##### puppetfile
+
+When Puppetfile content for a given environment is deployed, recursively
```
Comment (andersonmills) [inline doc/dynamic-environments/configuration.mkd]: same. Thumbs up for parallelism!

File: doc/dynamic-environments/configuration.mkd [hunk 6]
```diff
@@ -200,12 +200,81 @@ sources:
...
+is not also declared in that Puppetfile. Directories considered to be
+mananged by a Puppetfile include the default moduledir (typically "modules")
+as well as alternate directories specified as an "install\_path" option
+to any Puppetfile content declarations.
+
+#### purge\_whitelist
+
+The `purge_whitelist` setting exempts the specified filename patterns from
+being purged. This setting is currently only considered during `environment`
+level purging. (See above.) Must be a list of shell style filename patterns.
```
Comment (andersonmills) [inline doc/dynamic-environments/configuration.mkd]: "must" non-sentence again.

**Comment by scotje** [comment]
c7acddc should hopefully fix 1.9.3 and JRuby.

File: doc/dynamic-environments/configuration.mkd [hunk 7]
```diff
@@ -237,10 +237,16 @@ repo would no longer cause the matching environment to be purged.
 
 When a given environment is deployed, recursively remove any content found
 which is not either committed to the control repo branch that maps to that
-environment, or declared in a Puppetfile committed to that branch. Note that
-the .r10k-deploy.json file is exempt from this purging. Additionally, no
-content will be purged if an error is encountered while deploying the
-environment's Puppetfile content.
+environment, or declared in a Puppetfile committed to that branch.
```
Comment (andersonmills) [inline doc/dynamic-environments/configuration.mkd]: I think this "not... either... or..." is exactly why "neither... nor..." exists.

File: doc/dynamic-environments/configuration.mkd [hunk 8]
```diff
@@ -200,12 +200,92 @@ sources:
...
+encountered.
+
+Note that the .r10k-deploy.json file is exempt from this purging.
+
+##### puppetfile
+
+After Puppetfile content for a given environment is deployed, r10k will
+recursively remove any content found in a directory managed by the Puppetfile
+which is not also declared in that Puppetfile. Directories considered to be
+managed by a Puppetfile include the default moduledir (typically "modules")
```
Comment (andersonmills) [inline doc/dynamic-environments/configuration.mkd]: I think this should be "the specified moduledir (defaults to "modules" if not specified)"  Current wording is not clear to me.

**Comment by andersonmills** [comment]
:+1: on code. :+1: on docs except for default moduledir business in deploy-purge_levels-puppetfile section of configuration.

**Comment by scotje** [comment]
@andersonmills I rewrote that section a bit.

**Comment by andersonmills** [comment]
:+1: Fully and finally!

**Pull request merged**